031-custom_authentication.md

```diff
@@ -0,0 +1,104 @@
+
+This proposal focuses on supporting custom authentication in the Strimzi operator.
+
+## Current situation
+
+Currently, numerous authentication methods are [supported](https://github.com/strimzi/strimzi-kafka-operator/tree/0.25.0/api/src/main/java/io/strimzi/api/kafka/model/authentication) in Strimzi, and allows the user to configure which one to use. In addition, in the case a custom *authorizer* needs to be used, this is also currently [supported](https://github.com/strimzi/strimzi-kafka-operator/blob/0.25.0/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java#L541-L549). 
+
+However, there is no ability to specify custom authentication, which is what this proposal will focus on.
+
+## Motivation
+
+Supporting custom authentication allows greater flexibility for users of Strimzi who may have bespoke/proprietary authentication requirements. This would greatly broaden user adoption of the operator, as it would also support more common authentication schemes, such as [Amazon MSK](https://docs.aws.amazon.com/msk/latest/developerguide/security_iam_service-with-iam.html), and attract users who have much more stringent requirements (i.e. proprietary auth). 
+
+The “cost” of supporting such feature should be minimal as well, as the expectation for custom authentication passes the complexity onto the user. Strimzi should strictly support the passing of configuration values, and nothing else. It’s up to the user to ensure their authn/z schemes can be integrated with Strimzi. Considering this also requires custom built Strimzi Kafka images in order to load custom classes on to the classpath, it’s also expected that will only really apply to highly technical/pro users.
+
+## Proposal
+
+Add a CustomAuthenticationKafkaListener class, which supports an example of the following properties
+
+```
+listener.name.<listener-name>.oauthbearer.sasl.client.callback.handler.class=
+listener.name.<listener-name>.oauthbearer.sasl.server.callback.handler.class=
+listener.name.<listener-name>.oauthbearer.sasl.login.callback.handler.class=
+listener.name.<listener-name>.oauthbearer.connections.max.reauth.ms=
+listener.name.<listener-name>.sasl.enabled.mechanisms=
+listener.name.<listener-name>.oauthbearer.sasl.jaas.config=
+listener.name.<listener-name>.AWS_MSK_IAM.sasl.jaas.config=
+listener.name.<listener-name>.AWS_MSK_IAM.sasl.client.callback.handler.class=
+listener.security.protocol.map=
+principal.builder.class=
+```
+
+Really, this can be boiled down to: 
+
+```
+listener.name.<lister-name>.<auth-type>.*
+listener.name.<listener-name>.sasl.enabled.mechanisms
+security.protocol
+principal.builder.class
+```
+
+Taking this further, the desired yaml should look like: 
+
+```
+    listeners:
+      - name: tls
+        port: 9093
+        type: internal
+        tls: true
+        authentication:
+          type: custom
+          principal.builder.class: SimplePrincipal.class
+          security.protocol: SASL_SSL
```

Comment:
The `_SSL` part is driven by the `tls: true` flag on a higher level. So - assuming we support only SASL mechanisms - this has to be configured automatically in the code. If we would support also some _non-SASL_ mechanisms, it would need to be configurable with same `SASL` flag.

Reply:
Good point, my only follow up question is, should we support non-SASL based authentication workflows? To my understanding, there really wouldn't be a need for any, as mTLS should already be handled by a TLS listener? I suppose there is http basic authentication, which potentially wouldn't ever be compatible with Strimzi as the usage of certs for authentication.

Reply:
TBH, I don't know what other methods there would be. I guess it is only mTLS? Or is there something else? There were one or two requirements for a combination of mTLS and SASL in the past. If it is just mTLS, it might be maybe possible to add some more flag got that. If it is more than just mTLS, it might be easier to focus on SASL only.

Reply:
So mTLS should already be handled by `KafkaListenerAuthenticationTls`, as it enabled mTLS, plus from what I can tell, there are no additional properties a user would really need to modify.

Reply:
Yeah, but what I meant was that there are people interested in having mTLS and SASL in parallel on the same listener => you cannot combine your custom type authentication with `KafkaListenerAuthenticationTls` in a single listener (while that would be useful, it is one of the things where the API was not thought through properly and we missed this option and now it is hard to change).

Reply:
I'm most likely misunderstanding, so apologies for the back-and-forth, but is it possible to have both SASL + mTLS? Confluent docs mention (https://docs.confluent.io/platform/current/kafka/authentication_ssl.html): 
```
If any SASL authentication mechanisms are enabled for a given listener, then SSL client authentication is disabled–even if you have specified ssl.client.auth=required and the broker authenticates clients only using SASL on that listener.
```

Reply:
I don't know TBH. I was not aware of any such limitation - but that does not mean it is not there.

Reply:
[KIP-684](https://cwiki.apache.org/confluence/display/KAFKA/KIP-684+-+Support+mutual+TLS+authentication+on+SASL_SSL+listeners) will change this.

Reply:
Ah thanks for the link @tombentley ! Cool, then my proposal should definitely support this.